Source/WebGPU/WebGPU/Sampler.mm

```diff
@@ -272,41 +289,62 @@ static uint64_t floatToUint64(float f)
     return !!m_samplerIdentifier;
 }
 
+NSUInteger Sampler::aliveCount()
+{
+    Locker locker { samplerStateLock };
+    [weakSamplerStates _markNeedsCompaction];
+    [weakSamplerStates compact];
+    return weakSamplerStates.count;
+}
+
 id<MTLSamplerState> Sampler::samplerState() const
 {
     if (!m_samplerIdentifier)
         return nil;
 
-    Locker locker { samplerStateLock };
-    if (!cachedSamplerStates) {
-        cachedSamplerStates = [NSMutableDictionary dictionary];
-        lastAccessedKeys = [NSMutableOrderedSet orderedSet];
+    {
+        Locker locker { samplerStateLock };
+        if (!cachedSamplerStates) {
+            cachedSamplerStates = [NSMutableDictionary dictionary];
+            lastAccessedKeys = [NSMutableOrderedSet orderedSet];
+            weakSamplerStates = [NSPointerArray weakObjectsPointerArray];
+        }
+
+        id<MTLSamplerState> samplerState = [cachedSamplerStates objectForKey:m_samplerIdentifier];
+        if (samplerState) {
+            m_cachedSamplerState = samplerState;
+            return samplerState;
+        }
     }
 
-    id<MTLSamplerState> samplerState = [cachedSamplerStates objectForKey:m_samplerIdentifier];
-    if (samplerState)
-        return samplerState;
-
     id<MTLDevice> device = m_device->device();
     if (!device)
         return nil;
-    if (cachedSamplerStates.count >= device.maxArgumentBufferSamplerCount) {
+
+    if (aliveCount() >= device.maxArgumentBufferSamplerCount) {
+        Locker locker { samplerStateLock };
         if (!lastAccessedKeys.count)
             return nil;
 
         SamplerIdentifier* key = [lastAccessedKeys objectAtIndex:0];
         if (key)
             [cachedSamplerStates removeObjectForKey:key];
         [lastAccessedKeys removeObjectAtIndex:0];
-        ASSERT(cachedSamplerStates.count < device.maxArgumentBufferSamplerCount);
     }
 
-    samplerState = [device newSamplerStateWithDescriptor:createMetalDescriptorFromDescriptor(m_descriptor)];
+    if (aliveCount() >= device.maxArgumentBufferSamplerCount)
+        return nil;
+
+    id<MTLSamplerState> samplerState = [device newSamplerStateWithDescriptor:createMetalDescriptorFromDescriptor(m_descriptor)];
```

Comment:
If you get null from the system it means there is too many samplers, either yours or somebody else's.
What you can do is free your own -- you cannot free anybody elses'.
So you should error out once you're freed all your own and it still doesn't give out samplers.


I think the crucial thing to understand is that `aliveCount()` counts the "sampler objects WebGPU has created".
If you are going to create max amount of objects, that number really doesn't matter one bit. You just create the objects until you cannot.

It matters only if you are going to limit to something sane. E.g "Out of 2048 available unique sampler objects, WebGPU can create 20 and then error out".

Reply:
The concern with allowing `newSamplerStateWithDescriptor:` to return nil is the metal validation layer release asserts when the limit is exceeded. I would need to check the driver source what occurs when we exceed the global argument buffer process sampler limit with validation disabled to see if we can rely on this function returning nil.

Reply:
Ok, I see. It seems we need to track the created MTLSamplerStates with the weak ptrs.
Well, could we still fix the allocation as a loop and try to fix the mutexing to avoid the obvious races.